x/sync/manager.go

```diff
@@ -354,46 +400,252 @@ func (m *Manager) getAndApplyRangeProof(ctx context.Context, work *workItem) {
 		return
 	}
 
-	proof, err := m.config.Client.GetRangeProof(ctx,
-		&pb.SyncGetRangeProofRequest{
-			RootHash: targetRootID[:],
-			StartKey: &pb.MaybeBytes{
-				Value:     work.start.Value(),
-				IsNothing: work.start.IsNothing(),
-			},
-			EndKey: &pb.MaybeBytes{
-				Value:     work.end.Value(),
-				IsNothing: work.end.IsNothing(),
-			},
-			KeyLimit:   defaultRequestKeyLimit,
-			BytesLimit: defaultRequestByteSizeLimit,
+	request := &pb.SyncGetRangeProofRequest{
+		RootHash: targetRootID[:],
+		StartKey: &pb.MaybeBytes{
+			Value:     work.start.Value(),
+			IsNothing: work.start.IsNothing(),
 		},
-	)
+		EndKey: &pb.MaybeBytes{
+			Value:     work.end.Value(),
+			IsNothing: work.end.IsNothing(),
+		},
+		KeyLimit:   defaultRequestKeyLimit,
+		BytesLimit: defaultRequestByteSizeLimit,
+	}
+
+	requestBytes, err := proto.Marshal(request)
 	if err != nil {
+		m.finishWorkItem()
+		m.setError(err)
+		return
+	}
+
+	onResponse := func(ctx context.Context, _ ids.NodeID, responseBytes []byte, appErr error) {
+		defer m.finishWorkItem()
+
+		if err := m.handleRangeProofResponse(ctx, targetRootID, work, request, responseBytes, appErr); err != nil {
+			m.config.Log.Debug("dropping response", zap.Error(err))
+			m.retryWork(work)
+			return
+		}
+	}
+
+	if err := m.sendRequest(ctx, m.config.RangeProofClient, requestBytes, onResponse); err != nil {
+		m.finishWorkItem()
 		m.setError(err)
 		return
 	}
 
+	m.metrics.RequestMade()
+}
+
+func (m *Manager) sendRequest(ctx context.Context, client *p2p.Client, requestBytes []byte, onResponse p2p.AppResponseCallback) error {
+	if len(m.config.StateSyncNodes) == 0 {
+		return client.AppRequestAny(ctx, requestBytes, onResponse)
+	}
+
+	// Get the next nodeID to query using the [nodeIdx] offset.
+	// If we're out of nodes, loop back to 0.
+	// We do this try to query a different node each time if possible.
+	nodeIdx := atomic.AddUint32(&m.stateSyncNodeIdx, 1)
+	nodeID := m.config.StateSyncNodes[nodeIdx%uint32(len(m.config.StateSyncNodes))]
+	return client.AppRequest(ctx, set.Of(nodeID), requestBytes, onResponse)
+}
+
+func (m *Manager) retryWork(work *workItem) {
+	work.priority = retryPriority
+	work.queueTime = time.Now()
+	work.requestFailed()
+
+	m.workLock.Lock()
+	m.unprocessedWork.Insert(work)
+	m.workLock.Unlock()
+}
+
+// Returns an error if we should drop the response
+func (m *Manager) handleResponse(
+	bytesLimit uint32,
+	responseBytes []byte,
+	err error,
+) error {
```

Comment:
Can we make a better name for this function? This function doesn't actually handle the response... It returns an error if the caller shouldn't handle the response.